Add Ethereum network indexer (phase 1: blocks only)

Jannis opened the pull request:
I consider this generally ready for review!

This PR implements phase 1 of #297, including the following features:

1. A network indexer that indexes blocks from the selected network. (Transactions, logs, receipts, accounts, balances come in the next two phases.) This network indexer handles reorgs to an unlimited depth (bounded only by the memory of the machine the node runs on).

2. A `--network-subgraphs` CLI flag to enable network subgraph indexing per network, e.g. with
   ```sh
   graph-node ... --network-subgraphs ethereum/mainnet ethereum/kovan
   ```

3. A built-in GraphQL schema that allows Ethereum network subgraphs to be queried (or subscribed to) at `/subgraphs/ethereum/mainnet`. Apart from slightly different relationship fields, this schema is heavily based on Geth's [GraphQL schema](https://github.com/ethereum/go-ethereum/blob/master/graphql/schema.go).

4. Tests for basic indexing and reorg handling. These should be extended further by also verifying that the blocks are actually written to the store. Right now they only test the `Revert` and `AddBlock` events emitted by the indexer (which are only emitted after reverting/writing blocks successfully; however that doesn't mean the written data is correct).

## Database size

Regarding the size of the indexed data: The most recent 4000 mainnet blocks resulted in a Postgres database size of 20MB. That makes ~65GB for all 9,000,000 blocks, assuming they are the same size on average. They are probably smaller on average (older blocks were less busy), so we're looking at maybe 50GB just for the blocks (or their headers, rather).

## Review guide

I recommend reviewing commit by commit first. I've consolidated the PR into commits that mostly change only one thing at a time, so it's easier to follow what was added when.

It may help to read up on the terminology used across the network indexer here: https://github.com/graphprotocol/graph-node/blob/ec19ad21c7af971848643d2247ad355ead64ffb9/chain/ethereum/src/network_indexer/network_indexer.rs#L16-L54

The state machine for the network indexer is documented here: https://github.com/graphprotocol/graph-node/blob/ec19ad21c7af971848643d2247ad355ead64ffb9/chain/ethereum/src/network_indexer/network_indexer.rs#L509-L673

---

While this is being reviewed, I'll work on improving the tests to test data correctness and potentially generalizing the network indexer across chains. I've done some initial thinking to identify how the indexer depends on Ethereum right now and I think we can abstract that away.

leoyvens commented:
> potentially generalizing the network indexer across chains

This is probably quite a diff, would it be done in this PR or a follow up?

Jannis commented:
@leoyvens I'd be happy to make that a follow up, given the size of this PR.

Jannis commented:
One thing I'll still do is put metrics back in. I added the utilities (`Aggregate` and `.measure()`) to the PR and was using them at some point. But I rewrote this code about three times and dropped the metrics along the way.

Jannis commented:
I've added extensive instrumentation to enable Grafana dashboards like this one:

![image](https://user-images.githubusercontent.com/19324/69820282-11231300-1201-11ea-9b4e-7aa38826c6d5.png)

Jannis commented:
Tests generally pass, just sometimes they hang on Travis. I'm not sure yet what causes it.

Jannis commented on chain/ethereum/src/network_indexer/block_writer.rs:
```diff
@@ -0,0 +1,190 @@
+
+        Box::new(
+            // Add the block entity
+            self.set_entity(block.as_ref(), Some(vec![("isOmmer", false.into())]))
```
Agreed. I've changed this so ommer blocks are wrapped in an `Ommer` newtype and the `isOmmer` flag is now set in the `TryIntoEntity` implementations for `Ommer` and `BlockWithUncles`.

Jannis commented on chain/ethereum/src/network_indexer/mod.rs:
```diff
@@ -0,0 +1,102 @@
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct BlockWithUncles {
+    pub block: EthereumBlock,
+    pub uncles: Vec<Option<Block<H256>>>,
```
The docs are not super clear; I think https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_getUncleByBlockHashAndIndex and it's link to https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_getblockbyhash suggests that `None` means the uncle couldn't be found.

We can't allow different nodes to return different uncles though – we need them all for computing block rewards. So I think if an uncle (ommer) is not found, that's a serious reason to fail the network indexer. I'll drop the `Option`.

Jannis replied:
Updated

Jannis commented on chain/ethereum/src/network_indexer/mod.rs:
```diff
@@ -0,0 +1,102 @@
+    format_light_block(block.inner())
+}
+
+pub(crate) fn format_block_pointer(ptr: &EthereumBlockPointer) -> String {
```
Yeah, let's do it.

Jannis replied:
Done

Jannis commented on chain/ethereum/src/network_indexer/ethereum.graphql:
```diff
@@ -0,0 +1,86 @@
+  hash: Bytes!
+
+  """The parent block of this block."""
+  parent: Block
```
This is based on https://github.com/ethereum/go-ethereum/blob/d90d1db609c8d77baa422d49bd371207c06b4711/graphql/schema.go#L146-L147. Note that having a parent hash and having a parent block that _exists_ are two different things. Let's say genesis has `0x0...0` as its parent hash. That doesn't mean it has an actual parent.

Jannis replied:
We could store both the `parentHash` and the parent reference. What do you think?

Jannis replied:
The only block where that would be interesting is the genesis block though, I think. So it may be a lot of overhead for nothing. Because in all other cases, we can get to the parent block hash via `parent { hash }`.

leoyvens replied:
I see, yhea this really should be nullable.

Jannis commented on chain/ethereum/src/network_indexer/ethereum.graphql:
```diff
@@ -0,0 +1,86 @@
+  # logs: [Log!]!
+
+  """Size of the block in bytes."""
+  size: BigInt
```
I did this because it is optional here: https://tomusdrw.github.io/rust-web3/web3/types/struct.Block.html#structfield.size. I've seen on Etherscan that uncles don't report their size, so maybe that's why it is like that in rust-web3.

leoyvens replied:
Makes sense.

Jannis commented on chain/ethereum/src/network_indexer/convert.rs:
```diff
@@ -0,0 +1,158 @@
+}
+
+/// A value that can (maybe) be converted to an entity.
+pub trait TryIntoEntity {
```
Yeah, I'm not even sure this needs to be its own trait. `TryInto` should suffice.

Jannis replied:
There is another reason why these are local to the indexer though: Going forward (multi-blockchain), we'll have to pull a lot of the Ethereum types out of the `graph` crate. And then some of these traits and their implementations will have to live in the chain-specific code.

I'm still going to move some of this into `graph`.

Jannis replied:
Done

Jannis commented on chain/ethereum/src/network_indexer/network_indexer.rs:
```diff
@@ -0,0 +1,1540 @@
+                    );
+
+                    // Chain head wasn't new, try getting a new one.
+                    transition!(PollChainHead {
```
I thought so too initially; then I ran it and I noticed that — once it has caught up with the chain — it's only generating about 2-3 requests per second (because `eth_getBlockNumber("latest")` takes a few hundred milliseconds usually), which is not excessive. For chains that have block times of 5s or so I wouldn't want to add a delay greater than 500ms. But that would only reduce the requests per second from 2-3 to 1-2. Not sure it's worth it.

leoyvens replied:
Alright we can see how this works out in practice.

Jannis commented on chain/ethereum/src/network_indexer/network_indexer.rs:
```diff
@@ -0,0 +1,1540 @@
+type ChainHeadFuture = Box<dyn Future<Item = LightEthereumBlock, Error = Error> + Send>;
+type BlockPointerFuture = Box<dyn Future<Item = EthereumBlockPointer, Error = Error> + Send>;
+type BlockFuture = Box<dyn Future<Item = Option<BlockWithUncles>, Error = Error> + Send>;
+type BlockStream = Box<dyn Stream<Item = Option<BlockWithUncles>, Error = Error> + Send>;
```
It would be. How do I do that again? 😁

leoyvens replied:
Seeing that you already reset the indexer if this errors, I think yielding an error where you are yielding `None` should be enough.

Jannis replied:
Done! You're right, it's a good idea because it simplifies the remaining parts of the indexer.

Jannis commented on chain/ethereum/src/network_indexer/network_indexer.rs:
```diff
@@ -0,0 +1,1540 @@
+        //
+        // The assumption here is that maybe the network provider being
+        // used is temporarily serving from an outdated node.
+        let block_number = block.inner().number.unwrap().as_u64();
```
I _think_ you're right. Let me revisit this.

Jannis replied:
You _are_ right. I'm replacing this with an `except` just to be sure.

Jannis replied:
Done

Jannis commented on chain/ethereum/src/network_indexer/network_indexer.rs:
```diff
@@ -0,0 +1,1540 @@
+        // Check whether we have a reorg (parent of the new block != our local head).
+        if block.inner().parent_ptr() != state.local_head {
+            let depth = block.inner().number.unwrap().as_u64()
+                - state.local_head.map_or(0u64, |ptr| ptr.number);
```
Eh, you're right, this will never report the _real_ depth of the reorg. I do want that information, but I think I can only log it once we have found the common ancestor.

Jannis replied:
Updated

Jannis commented on chain/ethereum/src/network_indexer/network_indexer.rs:
```diff
@@ -0,0 +1,1540 @@
+                let state = state.take();
+
+                transition!(PollChainHead {
+                    local_head: state.old_local_head,
```
True, I like that.

Jannis replied:
Done

Jannis commented on graph/src/ext/futures.rs:
```diff
@@ -224,6 +228,10 @@ impl<F: Future> FutureExtension for F {
     }
+
+    fn measure<C: FnOnce(&Self::Item, Duration)>(self, callback: C) -> Measure<Self, C> {
+        Measure::new(self, callback)
```
I can move it into the indexer code.

Jannis replied:
And this is done also.

Jannis commented on chain/ethereum/src/network_indexer/mod.rs:
```diff
@@ -0,0 +1,102 @@
+    store: Arc<S>,
+    metrics_registry: Arc<dyn MetricsRegistry>,
+    start_block: Option<EthereumBlockPointer>,
+) -> impl Future<Item = NetworkIndexer, Error = ()>
```
Done, it's now a state in the indexer state machine.

Jannis commented:
@leoyvens I think I've addressed all comments with the appropriate changes. Could you take another look?

Jannis commented:
Rebased on top of master.

Jannis commented:
Reorg handling was simplified as per @leoyvens's suggestion. Reduced the implementation by about 600 lines and made it a lot easier to follow.

Jannis commented:
Left to do: Count consecutive reverts to capture and log reorg depths.

leoyvens commented on chain/ethereum/src/network_indexer/mod.rs:
```diff
@@ -0,0 +1,88 @@
+            difficulty: block.difficulty,
+            total_difficulty: block.total_difficulty,
+            seal_fields: block.seal_fields,
+            uncles: block.uncles,
```
Is correct and worth it to assert that this is empty?

Jannis replied:
Good question. I wouldn't think uncles ever report more uncles (that's not how it works). Asserting this might cause failures though. I'd rather have references to uncles of uncles in the resulting data that resolve to `null`.

leoyvens commented on chain/ethereum/src/network_indexer/network_indexer.rs:
```diff
@@ -0,0 +1,1196 @@
+                    "local_head" => state.local_head.map_or(
+                        String::from("none"), |ptr| format!("{}", ptr)
+                    ),
+                    "range" => format!("#{}..#{}", block_numbers.start, block_numbers.end-1),
```
Lets put the range in this log in square brackets to clarify that it's inclusive on both ends.

Jannis replied:
Done

leoyvens commented on chain/ethereum/src/network_indexer/network_indexer.rs:
```diff
@@ -0,0 +1,1196 @@
+            let event_sink = context.event_sink.clone();
+            let metrics_for_written_block = context.metrics.clone();
+
+            let section = { context.metrics.stopwatch.start_section("transact_block") };
```
These brackets are unecessary.

Jannis replied:
Done

lutter commented on chain/ethereum/src/network_indexer/ethereum.graphql:
```diff
@@ -0,0 +1,83 @@
+""" Block is an Ethereum block."""
+type Block @entity {
+  id: ID!
```
Before we start rolling this out, we should brush up the work I did on making it possible to make `ID` equivalent to `Bytes` for a subgraph so that these id's take only 20 instead of 40 (or 42) bytes to store. It will be transparent to the rest of the code, but requires that we pass a flag to `create_subgraph` that indicates whether `ID` should be a `String` or `Bytes`.

lutter replied:
The code for this is in `store::postgres::relational::Layout::new`, but it's not exposed to callers, and instead capped at `IdType::String` - we should expose this up in the callstack as arguments so it can be set in `create_subgraph` and is stored in the database (maybe as a field on `deployment_schemas`)

Jannis replied:
How much is there to do? And how much, if any, risk does that support introduce? Does it affect clients, filters, anything?

lutter replied:
Besides passing `IdType::Bytes` or `IdType::String` through when creating a subgraph, the following needs to be done in the `Store`:

* fix up a handful of places in `relational_queries.rs` where we assume that the id is a `String`
* Look at the type of the id column in `information_schema.columns` when starting a subgraph and decide whether it uses `Bytes` or `String` as the id

At the `Entity` layer, the block explorer would have to make sure to pass `id` as a `Value::Bytes` rather than a `Value::String`.

At the GraphQL layer, users have to pass the `id` as something that can be converted to `Value::Bytes`, and we'd have to do that conversion when coercing values. (I could make it so that we convert a `Value::String` into a `Value::Bytes` in `relational_queries.rs`, but that seems a bit hacky)

lutter replied:
I looked a little more, and to avoid changing too much in the code base, I think the best course of action is to keep that distinction within the relational mapping code. That means that code that deals with entity ID's continues to use strings, and the conversion from string to `bytea` and vice versa all happens in `relational_queries`. For users of the storage layer, the main change is that they might get a new error when the id is not a string in the form `0xdeadbeef`.

Those changes should be possible to do in a couple of days. The only change for the block explorer would be to pass `IdType::Bytes` when creating the schema.

Jannis replied:
Can we do this separately? We're not going to activate this feature right away.

lutter replied:
Yes, totally; we just need to do this before the feature goes live. We can migrate the database after the fact, but it's likely to take long (maybe hours) and during that time, the block explorer would be unavailable.

When this goes into a release, we need to make sure it's behind a feature flag so that users who install that release don't wind up with this data in their database which we would have to migrate.

lutter replied:
Opened #1414 to track this properly.

Jannis replied:
👍

Jannis replied:
It's already behind a `--network-subgraphs` CLI flag.

Jannis commented:
@lutter I've encountered one problem with this PR: with all of mainnet synced (~9M blocks), handling a reorg by reverting a single block took ~80s yesterday. It's done simply by calling `revert_block_operations` in https://github.com/graphprotocol/graph-node/blob/f3363e6a39d71c02e70927f4d33c7e5a4181f891/chain/ethereum/src/network_indexer/network_indexer.rs#L357-L361. Is there any way to speed this up?

lutter commented:
> @lutter I've encountered one problem with this PR: with all of mainnet synced (~9M blocks), handling a reorg by reverting a single block took ~80s yesterday. It's done simply by calling `revert_block_operations` in
> 
> https://github.com/graphprotocol/graph-node/blob/f3363e6a39d71c02e70927f4d33c7e5a4181f891/chain/ethereum/src/network_indexer/network_indexer.rs#L357-L361
> . Is there any way to speed this up?

I am sure there is ;) I'd have to look at a query plan.

lutter commented:
> > @lutter I've encountered one problem with this PR: with all of mainnet synced (~9M blocks), handling a reorg by reverting a single block took ~80s yesterday. It's done simply by calling `revert_block_operations` in
> > https://github.com/graphprotocol/graph-node/blob/f3363e6a39d71c02e70927f4d33c7e5a4181f891/chain/ethereum/src/network_indexer/network_indexer.rs#L357-L361
> > 
> > . Is there any way to speed this up?
> 
> I am sure there is ;) I'd have to look at a query plan.

I captured what we can do to improve this [here](https://github.com/graphprotocol/graph-node/issues/1415)

leoyvens commented:
The code looks good to me, I ran a smoke test and it worked fine. A question on the subgraph name, it's currently `ethereum/mainnet`. If we'll have a reserved namespace for these subgraphs, maybe it should be something more general, such as `chain/ethereum-mainnet` or `network/ethereum-mainnet`.

Jannis commented:
@leoyvens Yeah, that's a good point. I do like that the endpoint is `/subgraphs/name/ethereum/mainnet` right now, more than I'd like `/subgraphs/name/chain/ethereum-mainnet`.

One downside of using `ethereum/`, `substrate/`, `cosmos/` or whatever is that we'd have to own all these names in the network / GNS, instead of just having to own one prefix. Need to think about this a little bit.

Jannis commented:
I wonder if we can't do `/subgraphs/name/network/ethereum/mainnet`, `/subgraphs/name/network/substrate/polkadot` etc.

Zerim commented:
Was asked to weigh in on this... I agree with @leoyvens that it would be good to put the block explorer subgraphs in their own namespace, not only to free up those names in the common namespace but to indicate to Indexers that subgraphs under that namespace are "special" (i.e., don't use the standard WASM mapping code paths).

As far as having nice paths for our hosted service API endpoints, I'm in favor of making then compact, even if it doesn't perfectly follow the structure of normal subgraph API endpoints.

I.e, it could just be `/networks/ethereum/mainnet`

Jannis commented:
We can't easily serve subgraphs outside `/subgraphs` right now, but `/subgraphs/network{s}/ethereum/mainnet` would be possible. Let's do that.

Jannis commented:
@leoyvens @Zerim I've made it so that the following routes / subgraph names are used:

The subgraph name itself becomes `network/ethereum/mainnet`, `network/ethereum/kovan`.

The routes become: `/subgraphs/network/ethereum/mainnet` and `/subgraphs/network/ethereum/kovan`.
